blocks/randompruner.go

```diff
@@ -0,0 +1,238 @@
+package blocks
+
+import (
+	"bufio"
+	"context"
+	"errors"
+	"fmt"
+	"math/rand"
+	"os"
+	"path"
+	"sync"
+	"time"
+
+	"github.com/ipfs/go-cid"
+	flatfs "github.com/ipfs/go-ds-flatfs"
+	blockstore "github.com/ipfs/go-ipfs-blockstore"
+)
+
+type RandomPrunerConfig struct {
+	// The amount of bytes at which the blockstore will run a prune operation -
+	// cannot be zero
+	Threshold uint64
+
+	// Min bytes to remove during each prune operation - a value greater than or
+	// equal to MaxCapacityBytes shall prune all applicable blocks; a value of 0
+	// shall default to half of MaxCapacityBytes
+	PruneBytes uint64
+
+	// How long a block should remain pinned after it is used, before it is
+	// considered for pruning; defaults to one day
+	PinDuration time.Duration
+}
+
+// RandomPruner is a blockstore wrapper which removes blocks at random when disk
+// space is exhausted
+type RandomPruner struct {
+	blockstore.Blockstore
+	datastore   *flatfs.Datastore
+	threshold   uint64
+	pruneBytes  uint64
+	pinDuration time.Duration
+
+	// A list of "hot" CIDs which should not be deleted, and when they were last
+	// used
+	pins map[cid.Cid]time.Time
+
+	lk sync.Mutex
+}
+
+// The datastore that was used to create the blockstore is a required parameter
+// used for calculating remaining disk space - the Blockstore interface itself
+// does not provide this information
+func NewRandomPruner(inner blockstore.Blockstore, datastore *flatfs.Datastore, cfg RandomPrunerConfig) *RandomPruner {
+	if cfg.Threshold == 0 {
+		// Always, immediately fail for this config mistake - should there be a
+		// default?
+		log.Warnf("prune threshold is zero - this could potentially be a performance concern")
+	}
+
+	if cfg.PruneBytes == 0 {
+		cfg.PruneBytes = cfg.Threshold / 2
+	}
+
+	if cfg.Threshold > cfg.PruneBytes {
+		cfg.PruneBytes = cfg.Threshold
+	}
+
+	if cfg.PinDuration == 0 {
+		cfg.PinDuration = time.Hour * 24
+	}
+
+	return &RandomPruner{
+		Blockstore:  inner,
+		datastore:   datastore,
+		threshold:   cfg.Threshold,
+		pruneBytes:  cfg.PruneBytes,
+		pinDuration: cfg.PinDuration,
+	}
+}
+
+func (pruner *RandomPruner) Put(ctx context.Context, block Block) error {
+	pruner.updatePin(block.Cid())
+	pruner.Poll(ctx)
+	return pruner.Blockstore.Put(ctx, block)
+}
+
+func (pruner *RandomPruner) PutMany(ctx context.Context, blocks []Block) error {
+	for _, block := range blocks {
+		pruner.updatePin(block.Cid())
+	}
+	pruner.Poll(ctx)
+	return pruner.Blockstore.PutMany(ctx, blocks)
+}
+
+// Checks remaining blockstore capacity and prunes if maximum capacity is hit
+func (pruner *RandomPruner) Poll(ctx context.Context) {
+	pruner.cleanPins(ctx)
+
+	size, err := pruner.datastore.DiskUsage(ctx)
```

Comment:
checking on every operation may be suboptimal. might consider only doing this statistically, e.g. increment every time and actually check once/thousand, or make a call to random()

Reply:
oh you're right, could just store size on startup and increase it accordingly for each block that gets put. will add that.